File: App/LiveKitService.swift
import Foundation
import SwiftUI
import AVFoundation

enum LiveKitConnectionState: String {
    case disconnected, connecting, connected, reconnecting, error
}

struct ChatMessage: Identifiable, Equatable {
    let id: UUID
    let sender: String
    let body: String
    let sentAt: Date
    let isLocal: Bool
    let isVoiceNote: Bool
    var voiceNoteURL: URL?

    init(id: UUID = UUID(), sender: String, body: String, sentAt: Date = Date(), isLocal: Bool, isVoiceNote: Bool = false, voiceNoteURL: URL? = nil) {
        self.id = id
        self.sender = sender
        self.body = body
        self.sentAt = sentAt
        self.isLocal = isLocal
        self.isVoiceNote = isVoiceNote
        self.voiceNoteURL = voiceNoteURL
    }
}

/// Protocol-based LiveKit façade — UI works with stubs; SPM LiveKit client plugs in behind `#if canImport(LiveKit)`.
@MainActor
final class LiveKitService: ObservableObject {
    @Published private(set) var connectionState: LiveKitConnectionState = .disconnected
    @Published private(set) var roomName = AppConfig.defaultLiveKitRoom
    @Published private(set) var participants: [String] = []
    @Published var messages: [ChatMessage] = []
    @Published var micEnabled = true
    @Published var cameraEnabled = false
    @Published var statusMessage = ""
    @Published var callElapsedSeconds = 0
    @Published var isInCall = false
    @Published var isVideoCall = false

    private var ticker: Timer?
    private var audioRecorder: AVAudioRecorder?
    private var audioPlayer: AVAudioPlayer?
    @Published var isRecordingVoiceNote = false
    @Published var voiceNoteDraftURL: URL?

    var isConfigured: Bool { AppConfig.isLiveKitConfigured }

    func connect(room: String = AppConfig.defaultLiveKitRoom, asVideo: Bool = false) async {
        guard isConfigured else {
            statusMessage = "LiveKit URL missing. Add LIVEKIT_URL to Secrets.local.swift."
            connectionState = .error
            AppHaptics.error()
            return
        }
        roomName = room
        isVideoCall = asVideo
        cameraEnabled = asVideo
        connectionState = .connecting
        statusMessage = "Connecting to \(room)…"
        AppHaptics.medium()

        // #if canImport(LiveKit)
        // let room = Room(); try await room.connect(url: AppConfig.liveKitURL, token: token)
        // #endif

        try? await Task.sleep(for: .milliseconds(700))
        connectionState = .connected
        isInCall = true
        participants = ["You", "Studio"]
        statusMessage = asVideo ? "Video call connected" : "Voice call connected"
        startTicker()
        AppHaptics.success()
        appendSystem("Joined room \(room)")
    }

    func disconnect() {
        stopTicker()
        connectionState = .disconnected
        isInCall = false
        cameraEnabled = false
        participants = []
        callElapsedSeconds = 0
        statusMessage = "Call ended"
        AppHaptics.soft()
        appendSystem("Left the call")
    }

    func toggleMic() {
        micEnabled.toggle()
        AppHaptics.selection()
        statusMessage = micEnabled ? "Microphone on" : "Microphone muted"
    }

    func toggleCamera() {
        cameraEnabled.toggle()
        AppHaptics.selection()
        statusMessage = cameraEnabled ? "Camera on" : "Camera off"
    }

    func sendText(_ text: String, as sender: String = "You") {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        messages.append(ChatMessage(sender: sender, body: trimmed, isLocal: true))
        AppHaptics.light()
        // LiveKit data channel / text stream publishes here in production.
    }

    func startVoiceNote() {
        let url = FileManager.default.temporaryDirectory.appendingPathComponent("vn-\(UUID().uuidString).m4a")
        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
        ]
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker])
            try session.setActive(true)
            audioRecorder = try AVAudioRecorder(url: url, settings: settings)
            audioRecorder?.record()
            voiceNoteDraftURL = url
            isRecordingVoiceNote = true
            AppHaptics.rigid()
            statusMessage = "Recording voice note"
        } catch {
            statusMessage = "Could not start voice note: \(error.localizedDescription)"
            AppHaptics.error()
        }
    }

    func stopVoiceNoteAndSend(as sender: String = "You") {
        audioRecorder?.stop()
        isRecordingVoiceNote = false
        guard let url = voiceNoteDraftURL else { return }
        messages.append(ChatMessage(sender: sender, body: "Voice note", isLocal: true, isVoiceNote: true, voiceNoteURL: url))
        voiceNoteDraftURL = nil
        statusMessage = "Voice note sent"
        AppHaptics.success()
    }

    func playVoiceNote(_ url: URL) {
        do {
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.play()
            AppHaptics.light()
        } catch {
            statusMessage = "Could not play voice note"
            AppHaptics.error()
        }
    }

    private func appendSystem(_ body: String) {
        messages.append(ChatMessage(sender: "System", body: body, isLocal: false))
    }

    private func startTicker() {
        stopTicker()
        callElapsedSeconds = 0
        ticker = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.callElapsedSeconds += 1 }
        }
    }

    private func stopTicker() {
        ticker?.invalidate()
        ticker = nil
    }

    var callTimeLabel: String {
        let m = callElapsedSeconds / 60
        let s = callElapsedSeconds % 60
        return String(format: "%02d:%02d", m, s)
    }
}

File: Android/app/src/main/java/net/mrblindbandit/app/connect/LiveKitService.kt
package net.mrblindbandit.app.connect

import android.content.Context
import android.media.MediaPlayer
import android.media.MediaRecorder
import android.os.Build
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.delay
import net.mrblindbandit.app.config.AppConfig
import java.io.File
import java.util.UUID

data class ChatMessage(
    val id: String = UUID.randomUUID().toString(),
    val sender: String,
    val body: String,
    val isLocal: Boolean,
    val isVoiceNote: Boolean = false,
    val voiceNotePath: String? = null
)

enum class ConnectionState { Disconnected, Connecting, Connected, Error }

/** LiveKit façade — UI works with scaffold; production uses io.livekit:livekit-android. */
class LiveKitService(private val context: Context) {
    var connectionState by mutableStateOf(ConnectionState.Disconnected)
        private set
    var isInCall by mutableStateOf(false)
        private set
    var isVideoCall by mutableStateOf(false)
        private set
    var micEnabled by mutableStateOf(true)
    var cameraEnabled by mutableStateOf(false)
    var statusMessage by mutableStateOf("")
    var callElapsedSeconds by mutableIntStateOf(0)
        private set
    var isRecordingVoiceNote by mutableStateOf(false)
        private set
    val messages = mutableStateListOf<ChatMessage>()
    val participants = mutableStateListOf<String>()

    private var recorder: MediaRecorder? = null
    private var voiceNotePath: String? = null
    private var player: MediaPlayer? = null

    val isConfigured: Boolean get() = AppConfig.isLiveKitConfigured
    val callTimeLabel: String
        get() = "%02d:%02d".format(callElapsedSeconds / 60, callElapsedSeconds % 60)

    suspend fun connect(asVideo: Boolean) {
        if (!isConfigured) {
            connectionState = ConnectionState.Error
            statusMessage = "LiveKit URL missing. Set LIVEKIT_URL in local.properties."
            return
        }
        isVideoCall = asVideo
        cameraEnabled = asVideo
        connectionState = ConnectionState.Connecting
        statusMessage = "Connecting…"
        delay(700)
        // Production: Room.connect(url, token from server using Clerk session)
        connectionState = ConnectionState.Connected
        isInCall = true
        participants.clear()
        participants.addAll(listOf("You", "Studio"))
        statusMessage = if (asVideo) "Video call connected" else "Voice call connected"
        messages.add(ChatMessage(sender = "System", body = "Joined room ${AppConfig.DEFAULT_LIVEKIT_ROOM}", isLocal = false))
    }

    fun tick() {
        if (isInCall) callElapsedSeconds += 1
    }

    fun disconnect() {
        isInCall = false
        cameraEnabled = false
        connectionState = ConnectionState.Disconnected
        participants.clear()
        callElapsedSeconds = 0
        statusMessage = "Call ended"
        messages.add(ChatMessage(sender = "System", body = "Left the call", isLocal = false))
    }

    fun toggleMic() {
        micEnabled = !micEnabled
        statusMessage = if (micEnabled) "Microphone on" else "Microphone muted"
    }

    fun toggleCamera() {
        cameraEnabled = !cameraEnabled
        statusMessage = if (cameraEnabled) "Camera on" else "Camera off"
    }

    fun sendText(text: String) {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return
        messages.add(ChatMessage(sender = "You", body = trimmed, isLocal = true))
    }

    fun startVoiceNote() {
        val file = File(context.cacheDir, "vn-${UUID.randomUUID()}.m4a")
        try {
            @Suppress("DEPRECATION")
            recorder = if (Build.VERSION.SDK_INT >= 31) MediaRecorder(context) else MediaRecorder()
            recorder?.apply {
                setAudioSource(MediaRecorder.AudioSource.MIC)
                setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
                setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
                setOutputFile(file.absolutePath)
                prepare()
                start()
            }
            voiceNotePath = file.absolutePath
            isRecordingVoiceNote = true
            statusMessage = "Recording voice note"
        } catch (e: Exception) {
            statusMessage = "Could not start voice note: ${e.message}"
            isRecordingVoiceNote = false
        }
    }

    fun stopVoiceNoteAndSend() {
        try { recorder?.stop() } catch (_: Exception) {}
        recorder?.release()
        recorder = null
        isRecordingVoiceNote = false
        val path = voiceNotePath ?: return
        messages.add(ChatMessage(sender = "You", body = "Voice note", isLocal = true, isVoiceNote = true, voiceNotePath = path))
        voiceNotePath = null
        statusMessage = "Voice note sent"
    }

    fun playVoiceNote(path: String) {
        try {
            player?.release()
            player = MediaPlayer().apply {
                setDataSource(path)
                prepare()
                start()
            }
        } catch (e: Exception) {
            statusMessage = "Could not play voice note"
        }
    }
}
